### PR TITLE
Added spike archiving and support for stdp-synapses to precise neurons

### LINKED ISSUE #85: Precise models don't work with stdp_synapse?
The following script:
```
import nest

cell1 = nest.Create('iaf_neuron')
cell2 = nest.Create('iaf_psc_exp_ps')

conn_dict = {"rule": "all_to_all"}
syn_dict = {"model": "stdp_synapse"}
nest.Connect(cell1, cell2, conn_dict, syn_dict)
```
gives
```
Traceback (most recent call last):
  File "tmp.py", line 8, in <module>
    nest.Connect(cell1, cell2, conn_dict, syn_dict)
  File "....lib/python2.7/site-packages/nest/hl_api.py", line 153, in stack_checker_func
    return f(*args, **kwargs)
  File "..../lib/python2.7/site-packages/nest/hl_api.py", line 1123, in Connect
    sr('Connect')
  File "....lib/python2.7/site-packages/nest/__init__.py", line 81, in catching_sli_run
    raise hl_api.NESTError("{0} in {1}{2}".format(errorname, commandname, message))
pynestkernel.NESTError: IllegalConnection in Connect_g_g_D_D: Creation of connection is not possible.
```
If I replace `'iaf_psc_exp_ps` with `'iaf_psc_exp`, it works fine. 

`nest.GetDefaults('iaf_psc_exp_ps')` shows no `tau_minus` parameter, and the same is true for some other precise models I looked at.

Is there a reason why `iaf_psc_exp_ps` cannot work with `stdp_synapse`? If so, please could this be documented (unless I missed the documentation). If not, please could this be implemented?

### PR DESCRIPTION
**Pull request by heplesser**
This should fix #85.

- Spike times are archived as actual spike time, i.e., stamp at end of interval minus offset.
- This commit implies on semantic change: for precise neurons, `t_spike` now is the actual spike time, not the end of the time step
- The `offset` is no longer returned by `GetStatus` on precise models.
- Since both `t_spike` and `offset` are undocumented properties, this should not be much of a problem.
- Two tests adapted and one regression test added.

Suggested reviewers: @suku248 @abigailm I'd appreciate if you could review this one quickly, since my fix for #77 depends on it (in particular on parrot_neuron_ps).

### CODE REVIEW

File: precise/parrot_neuron_ps.h [hunk 1]
```diff
@@ -34,38 +34,32 @@ Name: parrot_neuron_ps - Neuron that repeats incoming spikes handling
...
+- Weights on connection to the parrot_neuron are ignored.
+- Weights on connections from the parrot_neuron are handled as usual.
+- Delays are honored on incoming and outgoing connections.
+
+Only spikes arriving on connections to port 0 will be repeated.
+Connections onto port 1 will be accepted, but spikes incoming
+through port 1 will be ignored. This allows setting exact pre-
+and post-synaptic spike times for STDP protocols by connecting
+two parrot neurons spiking at desired times by, e.g., a
+stdp_synapse onto port 1 on the post-synaptic parrot neuron.
```
Comment (tammoippen) [inline precise/parrot_neuron_ps.h]: I believe, this is not true, yet. You changed `parrot_neuron_ps::handles_test_event(...)` to accept connections to port 1, but `parrot_neuron_ps::handle(...)` is unchanged, adding all incoming events, not differentiating between ports.

**Comment by tammoippen** [comment]
Nice work! I have not looked at the sli code in detail, though. You get my 'thumbs up', when the points with the `parrot_neuron_ps` are addressed.

**Comment by heplesser** [comment]
@tammoippen Good catch! I now added a proper test and fixed some more issues. `parrot_neuron_ps` now passes the same set of tests as `parrot_neuron` wrt handling rports, multiplicity and plasticity of synpases between parrots.

**Comment by heplesser** [comment]
To test that spike multiplicity is handled properly in conjunction with STDP, I have now added a test that compares weight changes in an stdp_synapse between two parrot neurons and two precise parrot neurons, when multiple spikes occur during a single time step. This is based on an idea in #77, but is included in the PR for this issue, since it depends on rport 1 support in parrot_neuron_ps. The logic of the test is described in detail in the docstring of the test class.

Thanks, @flinz, for coming up with the idea of placing plastic synapses between parrot neurons for testing!

**Comment by flinz** [comment]
@heplesser sure, I'm happy its of use!

**Comment by tammoippen** [comment]
:+1: from my side!

File: precise/iaf_psc_exp_ps.cpp [hunk 2]
```diff
@@ -458,6 +452,7 @@ nest::iaf_psc_exp_ps::emit_spike_( const Time& origin,
   S_.is_refractory_ = true;
 
   // send spike
+  set_spiketime( Time::step( S_.last_spike_step_ ), S_.last_spike_offset_ );
   SpikeEvent se;
 
   se.set_offset( spike_offset );
```
Comment (krishnanj) [inline precise/iaf_psc_exp_ps.cpp]: se.set_offset( S_.last_spike_offset_ );

Comment (krishnanj) [inline precise/iaf_psc_exp_ps.cpp]: Only a minor change I think, otherwise :thumbsup:

**Comment by heplesser** [comment]
@krishnanj Thanks for your review, I fixed this now by completely eliminating the local variable.

**Comment by heplesser** [comment]
This PR should be merged after #160.

**Comment by krishnanj** [comment]
:+1:

**Comment by abigailm** [comment]
:+1:  will merge once #160 is merged

**Comment by heplesser** [comment]
@abigailm Thanks for your review! I see that this PR now has conflicts, possibly due to the merge of #160, or #171. I will explore this later today, got a plane to catch now.

**Comment by tammoippen** [comment]
Very nice work! Merging.

**Pull request merged**